crates/kv-store/src/sstable.rs

```diff
@@ -0,0 +1,1026 @@
+                if end_idx == table_idx {
+                    iter.back_to_key(end);
+                    // if the next back is invalid, the next should also be invalid
+                    if !iter.has_next_back() {
+                        iter.next();
+                    }
+                    (end_idx, None, None)
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_back_to_key(block, end);
+                    (end_idx, Some(iter), None)
+                }
+            }
+            Bound::Excluded(end) => {
+                let end_idx = table.find_back_block_idx(end);
+                if end_idx == table_idx {
+                    iter.back_to_key(end);
+                    // if the next back is invalid, the next should also be invalid
+                    if !iter.has_next_back() {
+                        iter.next();
+                    }
+                    (end_idx, None, Some(end))
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_back_to_key(block, end);
+                    (end_idx, Some(iter), Some(end))
+                }
+            }
+            Bound::Unbounded => {
+                let end_idx = table.meta.len() - 1;
+                if end_idx == table_idx {
+                    (end_idx, None, None)
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_seek_to_first(block);
+                    (end_idx, Some(iter), None)
+                }
+            }
+        };
+
+        let mut ans = if let Some(end_iter) = end_iter {
+            debug_assert!(end_idx > table_idx);
+            SsTableIter {
+                table,
+                next_block_iter: iter,
+                next_block_idx: table_idx,
+                back_block_iter: end_iter,
+                back_block_idx: end_idx as isize,
+                next_first: false,
+            }
+        } else {
+            debug_assert!(end_idx == table_idx);
+            SsTableIter {
+                table,
+                next_block_iter: iter.clone(),
+                next_block_idx: table_idx,
+                back_block_iter: iter,
+                back_block_idx: end_idx as isize,
+                next_first: true,
+            }
+        };
+        // the current iter is empty, but has next iter. we need to skip the empty iter
+        ans.skip_next_empty();
+        ans.skip_next_back_empty();
+
+        if let Some(key) = excluded {
+            if ans.has_next() && ans.next_key().unwrap() == key {
+                ans.next();
+            }
+        }
+        if let Some(key) = end_excluded {
+            if ans.has_next_back() && ans.next_back_key().unwrap() == key {
+                ans.next_back();
+            }
+        }
+
+        // need to skip empty block
```

Comment:
Why is this not handled by `ans.next()` or `ans.next_back()`

Reply:
This serves `peek_xxx()`, we need to make sure the iterator is *healthy* before calling `next()` `next_back()`

Reply:
How does it become unhealthy?

Reply:
For example, the current next iterator is empty due to `exclude` and we have more blocks that can be iterated. 
Now it's unhealthy, `has_next()` will return `false`, but we have another next Block to iter.